src/coprocessor/dag/storage/scanner.rs

```diff
@@ -0,0 +1,567 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::range::*;
+use super::ranges_iter::*;
+use super::{OwnedKvPair, Storage};
+use crate::coprocessor::Error;
+
+const KEY_BUFFER_CAPACITY: usize = 64;
+
+/// A scanner that scans over multiple ranges. Each range can be a point range containing only
+/// one row, or an interval range containing multiple rows.
+pub struct RangesScanner<T> {
+    storage: T,
+    ranges_iter: RangesIterator,
+
+    scan_backward_in_range: bool,
+    is_key_only: bool,
+
+    scanned_rows_per_range: Vec<usize>,
+
+    // The following fields are only used for calculating scanned range. Scanned range is only
+    // useful in streaming mode, where the client need to know the underlying physical data range
+    // of each response slice, so that partial retry can be non-overlapping.
+    is_scanned_range_aware: bool,
+    current_range_lower_key: Vec<u8>,
+    current_range_upper_key: Vec<u8>,
+    working_range_begin_key: Vec<u8>,
+    working_range_end_key: Vec<u8>,
+}
+
+impl<T: Storage> RangesScanner<T> {
+    pub fn new(
+        storage: T,
+        ranges: Vec<Range>,
+        scan_backward_in_range: bool, // TODO: This can be const generics
```

Comment:
Can we define a struct and assign a name to each boolean `(false, false, false) / (true, true, false), ...blabla` is really hard to remember the meaning of each paramater)?

Reply:
Once we use const generics it should be a problem any more.

Reply:
There is a long time gap to wait for the const generics GA.